### PR TITLE
[EEG] Add view/edit functionality and permissions for EEG BIDS derivatives

### PR DESCRIPTION
**Pull request by AlexandraLivadas**
1. Added a new permission for a user to view and/or edit annotations for specific entries in the electrophysiology browser module. 
2. Created functions in the backend of the electrophysiology browser module to get and display derivative data from the database and to write derivative data to the database. There is no view/edit/delete functionality in the front-end added by this PR.
3. Added the ability to download annotation files for a given session. Created a function that updates the annotation files to match changes in the database before they are downloaded by a user. This function can be called by a cronjob using the `tools/update_annotation_files.php` to update all the files periodically.

**Note:** The `tools/updateDerivativesTest` was added for testing purposes and should be removed before merging.  

#### Testing instructions

To test the SQL patch: 
1. Source the `2021-04-23-annotation_permissions.sql` patch or refresh the database with the edited SQL permissions schema and see that the correct entries are added to the database.

To test the download functionality:
1. Check that your database includes some derivative datasets and files. 
2. Navigate to the 
3. See that the button to download annotation files is visible and clickable
4. Check that clicking this button correctly downloads a .tgz file with the correct annotation files
5. Add new data to the `physiological_annotation_instance` and `physiological_annotation_parameter` tables for the respective files. 
6. Re-download the files and see that they are correctly updated with the new information.

Related to #7345

### CODE REVIEW

File: modules/electrophysiology_browser/php/sessions.class.inc
Comment (laemtl) [inline modules/electrophysiology_browser/php/sessions.class.inc:65]: I think you're missing the superuser case :)

File: modules/electrophysiology_browser/php/sessions.class.inc
Comment (laemtl) [inline modules/electrophysiology_browser/php/sessions.class.inc:78]: Same here!

**Review by laemtl — CHANGES_REQUESTED** [review]

File: modules/electrophysiology_browser/php/sessions.class.inc
Comment (AlexandraLivadas) [reply modules/electrophysiology_browser/php/sessions.class.inc:65]: I think the superuser case is handled in the `hasPermission` method at [this line](https://github.com/aces/Loris/blob/2fceb83b8f11bc2b315e279d260730e3d11a40a5/php/libraries/UserPermissions.class.inc#L110)

Comment (laemtl) [reply modules/electrophysiology_browser/php/sessions.class.inc:65]: True! My mistake, it's all good then.

**Review by laemtl — APPROVED** [review]

**Comment by christinerogers** [comment]
Hi Alex, To align with the Imaging Browser permissions, I'd suggest just adding one permission to Add annotations. 
The capacity to Edit/Delete annotations can be tackled after the user can first Add annotations (top priority) 

cc @cmadjar any preferences here?

**Comment by cmadjar** [comment]
@christinerogers I may be wrong but I think on the imaging side, when a person can do QC, they can edit later on the QC values as well. Any reason why there should be a separation between add/edit/delete for EEG?

**Comment by christinerogers** [comment]
> @christinerogers I may be wrong but I think on the imaging side, when a person can do QC, they can edit later on the QC values as well. Any reason why there should be a separation between add/edit/delete for EEG?

Hi Cecile - agreed, they should be governed by the same permissions, I was just thinking that the functionality to edit an existing annotation will be added after the Add Annotation functionality (one step at a time).

**Comment by AlexandraLivadas** [comment]
@christinerogers I removed the `View` permissions (as those are covered by the pre-existing permissions for the module) but kept the `Create/Edit`permission as is. For now, I will only be writing the functionality to add new annotations and not to edit existing ones.

File: modules/electrophysiology_browser/README.md
Comment (christinerogers) [inline modules/electrophysiology_browser/README.md:34]: perfect. 👍

File: modules/electrophysiology_browser/README.md
Comment (christinerogers) [inline modules/electrophysiology_browser/README.md:26]:
```suggestion
permissions to add or modify annotations (derivatives) for data from all sites in this module.
```
@AlexandraLivadas  can you update these words `from all sites` if not accurate ?  does your permission to add/edit on all-sites or just the user's affiliated sites?

File: modules/electrophysiology_browser/README.md
Comment (christinerogers) [inline modules/electrophysiology_browser/README.md:20]:
```suggestion
datasets can also be accessed and annotated via this module.
```
annotated not edited.

File: modules/electrophysiology_browser/README.md
Comment (christinerogers) [inline modules/electrophysiology_browser/README.md:7]:
```suggestion
annotations (derivatives) for each recording.
```
updating for accuracy

File: modules/electrophysiology_browser/README.md
Comment (christinerogers) [inline modules/electrophysiology_browser/README.md:40]:
```suggestion
You can download all the files related to a recording (channel information,
electrode information, task event information, the actual recording) -- as well as its derivatives (annotations and their metadata).
```
updated this for clarity.  also no ellipses in documentation if you can help it...

**Review by christinerogers — CHANGES_REQUESTED** [review]
Alex - nice documentation update. 
Can you accept my changes and I left you one fact-check to followup on, inline below. 

More rainy day documetnation updates for this module will include: 
- [ ] Help text 
- [ ] updating when edit/delete actually become feasible.

File: modules/electrophysiology_browser/README.md
Comment (AlexandraLivadas) [reply modules/electrophysiology_browser/README.md:26]: I will update this because it does not apply for all-sites. The 2 permissions that existed before determine if the session can be viewed and this permission just determines if the annotations of the session being viewed can be edited.

File: modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js
Comment (laemtl) [inline modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:22]: For clarity, can this key and the props be renamed for something more explicit (like physiologicalFileID?)

File: modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js
Comment (laemtl) [inline modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:254]: handleClick is too vague maybe? I think something more descriptive could help (like annotationsUpdate).

File: modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js
Comment (laemtl) [inline modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:569]: Oups :) console.log can be removed.

File: modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js
Comment (laemtl) [inline modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:573]: I believe PUT is semantically incorrect: https://www.guru99.com/put-vs-post.html#5
This operation is not idempotent.

File: modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js
Comment (laemtl) [inline modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:585]: Not needed?

Comment (laemtl) [reply modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:585]: Do you have cases where an error can occur? For example, if the physiologicalID is not found.

File: modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js
Comment (laemtl) [inline modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:604]: Issue?

File: modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js
Comment (laemtl) [inline modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:605]: text can be a const.

File: modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js
Comment (laemtl) [inline modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:603]: For variables with different values, I think it's better to define here and assign either in the if or else case.

File: modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js
Comment (laemtl) [inline modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:624]: allowOutsideClick and allowEscapeKey are convenient if the user does not need to make a choice. Why setting them to false?

File: modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js
Comment (laemtl) [inline modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:641]: url should not have a default value I think.

File: modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js
Comment (laemtl) [inline modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:631]: Is fileId missing?

File: modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js
Comment (laemtl) [inline modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:635]: I think the standard name for this is action.

File: modules/electrophysiology_browser/jsx/electrophysiologySessionView.js
Comment (laemtl) [inline modules/electrophysiology_browser/jsx/electrophysiologySessionView.js:266]: Should we rename fileID for something more descriptive?

**Review by laemtl — CHANGES_REQUESTED** [review]

File: tools/updateDerivativesTest.php
Comment (christinerogers) [inline tools/updateDerivativesTest.php:62]: Hard coded

**Review by christinerogers — CHANGES_REQUESTED** [review]
Really nice to see this effort, thank you @AlexandraLivadas 
A few early comments on the script - requesting changes, see comments below. 
Also to do for this PR: 
- [ ] Add to the module readme: Downloads packages for this module (and for DQT e.g.) need regular verification/updating via a script called ___. To set up, configure, and run it (including via cronjob), here are the steps: ...   (roughly 2-4 sentences)

File: tools/updateDerivativesTest.php
Comment (AlexandraLivadas) [reply tools/updateDerivativesTest.php:62]: @christinerogers This is a test script I'm using to test the functionalities I added and will be removed before merging.

File: SQL/New_patches/2021-04-26_add_annotation_data.sql
Comment (ridz1208) [inline SQL/New_patches/2021-04-26_add_annotation_data.sql:8]: these drops need to be added to the 9999 schema

File: modules/electrophysiology_browser/php/sessions.class.inc
Comment (ridz1208) [reply modules/electrophysiology_browser/php/sessions.class.inc:67]: same

File: modules/electrophysiology_browser/php/sessions.class.inc
Comment (ridz1208) [inline modules/electrophysiology_browser/php/sessions.class.inc:842]: I'm not too involved in the EEG design usually so this comment might be inaccurate.

This function does a ton of SQL selects, some even in a loop of instances, which is a good way to waste a lot of time. Is there no way to compile everything that needs to be queried and then run a fraction of the queries to get all the needed data ?

File: modules/electrophysiology_browser/php/sessions.class.inc
Comment (ridz1208) [inline modules/electrophysiology_browser/php/sessions.class.inc:582]: you are building and running queries above this point which will have no effect (as far as i can see) if the user doesn't have the permission, that's inefficient.

You should consider checking for permission as early as possible and returning without wasting time

File: modules/electrophysiology_browser/php/sessions.class.inc
Comment (ridz1208) [inline modules/electrophysiology_browser/php/sessions.class.inc:828]: same comment as above, you don't need do do a lot of this if your user doesn't have the proper permission. do your permission check first

**Review by ridz1208 — CHANGES_REQUESTED** [review]
Requested changes on the SQL side to fit LORIS standards

On the code side, I only did  review for red flags as I would not be able to comment on annotations and EEG logic.

File: SQL/New_patches/2021-04-26_add_annotation_data.sql
Comment (AlexandraLivadas) [reply SQL/New_patches/2021-04-26_add_annotation_data.sql:8]: These were added in PR #7345. As mentioned in this PR description, this is a patch that I am using to add data to the database for testing purposes and it will be deleted before merging this PR.

File: modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js
Comment (AlexandraLivadas) [reply modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js]: Yes there are cases where the file is not found or the file path is incorrect/not found as well

File: modules/electrophysiology_browser/php/sessions.class.inc
Comment (AlexandraLivadas) [reply modules/electrophysiology_browser/php/sessions.class.inc:517]: @laemtl This method will be part of the initial GET request to view the annotations in the frontend (similar to `getDownloadLinks` above) so I followed the same formatting and made this method public. Let me know if this makes sense!

File: modules/electrophysiology_browser/php/models/electrophysioannotations.class.inc
Comment (laemtl) [inline modules/electrophysiology_browser/php/models/electrophysioannotations.class.inc:129]: Shouldn't this be $electrophysioFile?

File: modules/electrophysiology_browser/php/models/electrophysioannotations.class.inc
Comment (laemtl) [inline modules/electrophysiology_browser/php/models/electrophysioannotations.class.inc:130]: Shouldn't this be $user?

File: modules/electrophysiology_browser/php/models/electrophysioannotations.class.inc
Comment (laemtl) [inline modules/electrophysiology_browser/php/models/electrophysioannotations.class.inc:128]: We already have a description field below. Should this one be generic?

File: modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js
Comment (christinerogers) [inline modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:580]:
```suggestion
      let title = 'Your download will begin in a few moments';
      let text = 'Thank you for your patience';
```

Comment (christinerogers) [reply modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:580]: Hi @AlexandraLivadas  - Could you accept this change to the messaging ? thanks, 
Christine

**Review by christinerogers — CHANGES_REQUESTED** [review]
tweaks below.  
thanks for your efforts here.

File: modules/electrophysiology_browser/php/models/electrophysioannotations.class.inc
Comment (AlexandraLivadas) [reply modules/electrophysiology_browser/php/models/electrophysioannotations.class.inc:129]: No. In the BIDS specifications there are two fields: `Sources` and `IntendedFor`. IntendedFor is defined as: `REQUIRED. Link to the data file for which the annotations are intended.` and  Sources is defined as `RECOMMENDED. Relative link or description of the file(s) used to make the annotations.`. For the IntendedFor entry, I link this data to a physiological file ID in the database.

Comment (laemtl) [reply modules/electrophysiology_browser/php/models/electrophysioannotations.class.inc:129]: Following our discussion, I think 'Sources' is not necessary and 'this can be removed.

**Comment by cmadjar** [comment]
@AlexandraLivadas to work on this.

**Comment by christinerogers** [comment]
Hi @AlexandraLivadas - could you remind me, what's the status here?  just waiting for free time to work on this or a new maintainer - is that about the size of it?

**Comment by AlexandraLivadas** [comment]
> Hi @AlexandraLivadas - could you remind me, what's the status here? just waiting for free time to work on this or a new maintainer - is that about the size of it?

Hi @christinerogers. I will review right now but I think the only thing left here is to configure the cron job and then update the documentation with instructions.

**Review by christinerogers — CHANGES_REQUESTED** [review]
Hi Alex , 
I just added some comments on your latest commit (instead of on the whole PR) -
Let me know if you have any questions --

**Comment by AlexandraLivadas** [comment]
> "Updating File Download packages" i think is more accurate to say, perhaps ?

@christinerogers I don't think this would make it clear, since **all** derivative files are updated with the script, whereas only the files that are being downloaded are updated when the download button is pressed.

**Comment by christinerogers** [comment]
> > "Updating File Download packages" i think is more accurate to say, perhaps ?
> 
> @christinerogers I don't think this would make it clear, since **all** derivative files are updated with the script, whereas only the files that are being downloaded are updated when the download button is pressed.

Ok, spell these 2 different update processes out so it's clear for the module documentation.

**Comment by christinerogers** [comment]
> cut this info about what a cronjob is. assume the user knows.

@AlexandraLivadas I don't think this comment has been addressed.  Please cut all the info introducing what a cronjob is, what the settings mean, etc.

**Comment by christinerogers** [comment]
> > > "Updating File Download packages" i think is more accurate to say, perhaps ?
> > 
> > 
> > @christinerogers I don't think this would make it clear, since **all** derivative files are updated with the script, whereas only the files that are being downloaded are updated when the download button is pressed.
> 
> Ok, add just enough clarification to let an admin know how the file updates (can) happen in each case - so it's clear for the module documentation.

Has this comment been addressed?  I think you re-requested my review after my comment.

**Comment by AlexandraLivadas** [comment]
@christinerogers This is ready for re-review :)

File: modules/electrophysiology_browser/README.md
Comment (christinerogers) [inline modules/electrophysiology_browser/README.md:44]: proofreading / please correct: 
maded 
directly -> immediately
slected -> selected
downloading is triggered -> the actual download process starts
to simplify, Replace "All derivative files in the file system can be updated using the..." with "A script is provided to perform this [script name and link."

File: modules/electrophysiology_browser/README.md
Comment (christinerogers) [inline modules/electrophysiology_browser/README.md:46]: mmm

Comment (christinerogers) [reply modules/electrophysiology_browser/README.md:46]:
```suggestion
Below is a sample crontab entry to execute this script *every night* at midnight  
```
Please also change the cronjob fields so that it's not first of the month, but every night

**Review by christinerogers — CHANGES_REQUESTED** [review]
Minor updates and typo fixes, please and thank you :) 
slack me when it's ready for re-review 
C

**Review by christinerogers — CHANGES_REQUESTED** [review]
One change request you missed: 
> downloading is triggered -> the actual download process starts

and -- my apologies, please scratch the word "immediately" (no adverb needed there i don't think)

**Review by christinerogers — APPROVED** [review]

**Comment by AlexandraLivadas** [comment]
@ridz1208 @laemtl This is ready for re-review!

**Comment by christinerogers** [comment]
@AlexandraLivadas this PR is critical this week.  Could you rebase and let me know what next steps are? 

cc: @laemtl

**Review by christinerogers — APPROVED** [review]
Approving.  
@AlexandraLivadas are you still working on TestPlan updates for this PR per @cmadjar 's last review?

**Review by cmadjar — CHANGES_REQUESTED** [review]
@AlexandraLivadas the test plan still needs to be edited to add manual testing step for the new permission and the download capabilities of the annotation.

**Comment by AlexandraLivadas** [comment]
@cmadjar I have added instructions to test the download button for annotations. The permission I have added will only be relevant to test once there is frontend functionality for editing annotations. Let me know if the changes I made look good!

**Review by cmadjar — APPROVED** [review]
Thank you @AlexandraLivadas! LGTM.

File: modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js
Comment (christinerogers) [reply modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:100]: @AlexandraLivadas Is this comment outstanding or addressed / resolved?

File: SQL/New_patches/2021-02-19_electrophysiology_annotation_tables.sql
Comment (driusan) [inline SQL/New_patches/2021-02-19_electrophysiology_annotation_tables.sql:4]: These drop tables should not be in patches for existing projects. It adds an unnecessary risk of data loss.

**Pull request merged**
